vowpalwabbit/array_parameters.h

```diff
@@ -0,0 +1,181 @@
+#pragma once
+#include <stdio.h>
+#include <iostream>
+#include <stdlib.h>
+#include <string.h>
+
+#ifndef _WIN32
+#include <sys/mman.h>
+#endif
+
+using namespace std;
+typedef float weight;
+
+class weight_vector;
+
+class weights_iterator_iterator
+{
+private:
+	weight* _cur;
+public:
+	weights_iterator_iterator(weight* cur)
+		: _cur(cur)
+	{ }
+	
+	weight& operator*() { return *_cur; }
+
+	weights_iterator_iterator& operator++()
+	{
+		_cur++;
+		return *this;
+	}
+
+	template<typename T>
+	weights_iterator_iterator operator+(T index) { return weights_iterator_iterator(_cur + index); }
+
+	template<typename T>
+	weights_iterator_iterator& operator+=(T index)
+	{
+		_cur += index;
+		return *this;
+	}
+
+	weights_iterator_iterator& operator=(const weights_iterator_iterator& other)
+	{
+		_cur = other._cur;
+		return *this;
+	}
+
+	bool operator==(const weights_iterator_iterator& rhs) { return _cur == rhs._cur; }
+	bool operator!=(const weights_iterator_iterator& rhs) { return _cur != rhs._cur; }
+
+};
+class weights_iterator
+{
+private:
+	weight* _current;
+	uint32_t _stride_shift;
+
+public:
+	typedef weights_iterator_iterator w_iter;
+
+	weights_iterator(weight* current, uint32_t stride_shift)
+		: _current(current), _stride_shift(stride_shift)
+	{ }
+
+	weight& operator*() { return *_current; }
+
+	weights_iterator& operator++()
+	{
+		_current += _stride_shift;
+		return *this;
+	}
+
+	template<typename T>
+	weights_iterator operator+(T index) { return weights_iterator(_current + (index*_stride_shift), _stride_shift); }
+
+	template<typename T>
+	weights_iterator& operator+=(T index)
+	{
+		_current += (index*_stride_shift);
+		return *this;
+	}
+
+	weights_iterator& operator=(const weights_iterator& other)
+	{
+		_current = other._current;
+		_stride_shift = other._stride_shift;
+		return *this;
+	}
+
+	bool operator==(const weights_iterator& rhs) { return _current == rhs._current; }
+	bool operator!=(const weights_iterator& rhs) { return _current != rhs._current; }
+
+	//to iterate within a bucket
+	w_iter begin() { return w_iter(_current); }
+	w_iter end(size_t offset) { return w_iter(_current + offset); }
+
+};
+
+class weight_vector //different name? (previously array_parameters)
+{
+private:
+	weight* _begin;
+	uint64_t _weight_mask;  // (stride*(1 << num_bits) -1)
+	uint32_t _stride_shift;
+	size_t _size;
+	bool _seeded;
```

Comment:
Ok, makes sense.

-John

On 06/28/2016 02:00 PM, Aarti Bagul wrote:
>
> In vowpalwabbit/array_parameters.h 
> <https://github.com/JohnLangford/vowpal_wabbit/pull/1009#discussion_r68809246>:
>
> > +	bool operator!=(const weights_iterator& rhs) { return _current != rhs._current; }
> > +
> > +	//to iterate within a bucket
> > +	w_iter begin() { return w_iter(_current); }
> > +	w_iter end(size_t offset) { return w_iter(_current + offset); }
> > +
> > +};
> > +
> > +class weight_vector //different name? (previously array_parameters)
> > +{
> > +private:
> > +	weight* _begin;
> > +	uint64_t _weight_mask;  // (stride*(1 << num_bits) -1)
> > +	uint32_t _stride_shift;
> > +	size_t _size;
> > +	bool _seeded;
>
> I used _seeded in the destructor to make sure that the weight vector 
> is not freed if it is shared with another instance
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/JohnLangford/vowpal_wabbit/pull/1009/files/d792eb33cc19d5e7b7214f45686afcb37a31ad45#r68809246>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/AAE25k3Key-HlbBomQ94TDnxyMzzfvwSks5qQWElgaJpZM4Iwg44>.
>